pkg/util/watch.go

```diff
@@ -409,10 +359,72 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
 		}
 	}()
 
-	return &store, nil
+	return store, nil
+}
+
+type ProcessEventArgs[T any, P WatchObject[T]] struct {
+	Name   string
+	UID    types.UID
+	Config WatchConfig
+	Event  watch.Event
+	Object P
 }
 
-// WatchStore provides an interface for getting information about a list of Ts using the event
+func ProcessEvent[T any, P WatchObject[T]](
+	store *WatchStore[T],
+	handlers WatchHandlerFuncs[P],
+	args ProcessEventArgs[T, P],
+) error {
+	switch args.Event.Type {
+	case watch.Added:
+		store.mutex.Lock()
+		defer store.mutex.Unlock()
+
+		if _, ok := store.objects[args.UID]; ok {
+			return fmt.Errorf(
+				"watch %s: received add event for object %s that we already have",
+				args.Config.LogName, args.Name,
+			)
+		}
+		store.objects[args.UID] = (*T)(args.Object)
+		handlers.AddFunc((*T)(args.Object), false)
+	case watch.Bookmark:
+		// Nothing to do, just serves to give us a new ResourceVersion.
```

Comment:
Maybe worth having a similar "don't call this function with `watch.Error`" thing for `watch.Bookmark`; this is kind of just noise here, and we can filter it out at the callsite.

Reply:
I think there's merit in being exhaustive in switch statements like these. It's also the case that if we filter things out in the main loop, then this function becomes less useless as the stateless filter through which events flow.